integrations/GA/browser.js

```diff
@@ -29,88 +56,638 @@ class GA {
+        payload,
+        setCustomDimenionsAndMetrics(
+          rudderElement.message.properties,
+          this.inputs
+        )
+      );
+
+      ga("send", "event", payload);
+      logger.debug("in GoogleAnalyticsManager track");
+    }
   }
 
   page(rudderElement) {
+    var dimensionsArray = {};
+    for (let val of this.dimensions) {
+      dimensionsArray[val.from] = val.to;
+    }
+    var metricsArray = {};
+    for (let val of this.metrics) {
+      metricsArray[val.from] = val.to;
+    }
+    var contentGroupingsArray = {};
+    for (let val of this.contentGroupings) {
+      contentGroupingsArray[val.from] = val.to;
+    }
     logger.debug("in GoogleAnalyticsManager page");
-    var path =
-      rudderElement.message.properties && rudderElement.message.properties.path
-        ? rudderElement.message.properties.path
-        : undefined;
-    var title = rudderElement.message.properties && rudderElement.message.properties.title
-        ? rudderElement.message.properties.title
-        : undefined;
-    var location = rudderElement.message.properties && rudderElement.message.properties.url
-        ? rudderElement.message.properties.url
-        : undefined;
 
-    if (path) {
-      ga("rudder_ga.set", "page", path);
+    var category = rudderElement.message.properties.category;
+    var eventProperties = rudderElement.message.properties;
+    var name =
+      rudderElement.message.properties.category +
+      " " +
+      rudderElement.message.name;
+    var campaign = rudderElement.message.context.campaign | {};
+    var pageview = {};
+    var pagePath = path(eventProperties, this.includeSearch);
+    var pageReferrer = rudderElement.message.properties.referrer || "";
+    var pageTitle;
+    if (
+      !rudderElement.message.properties.category &&
+      !rudderElement.message.name
+    )
+      pageTitle = eventProperties.title;
+    else if (!rudderElement.message.properties.category)
+      pageTitle = rudderElement.message.name;
+    else if (!rudderElement.message.name)
+      pageTitle = rudderElement.message.properties.category;
+    else pageTitle = name;
+
+    pageview.page = pagePath;
+    pageview.title = pageTitle;
+    pageview.location = eventProperties.url;
+
+    if (campaign) {
+      if (campaign.name) pageview.campaignName = campaign.name;
+      if (campaign.source) pageview.campaignSource = campaign.source;
+      if (campaign.medium) pageview.campaignMedium = campaign.medium;
+      if (campaign.content) pageview.campaignContent = campaign.content;
+      if (campaign.term) pageview.campaignKeyword = campaign.term;
     }
 
-    if (title) {
-      ga("rudder_ga.set", "title", title);
+    // Reset custom dimension which are previously set.
+    // Uses the configured dimensions as:
+    // this.dimensions: { "fruit": "dimension1" }
+    // this.resetCustomDimensions: [ "fruit" ]
+    // --> resetCustomDimensions: { "dimension1": null }
+
+    var resetCustomDimensions = {};
```

Comment:
why do we need to reset?

Reply:
reset custom dimensions can be set by user in ui. whatever he sets that property will be reset.